File: twitchrolls.go
package main

import (
  "sort"
  "net/http"
  "time"
  "strings"
  "math/rand"
  "github.com/valyala/fasthttp"
  "github.com/json-iterator/go"
  "fmt"
  "io/ioutil"
  "strconv"
  "bytes"
  "net/url"
  "errors"
)

var json = jsoniter.ConfigCompatibleWithStandardLibrary

type APIPagination struct {
  Cursor string
}
type StreamInfoFake struct {
  Viewer_count int
  Game_id string
  User_name string
  Thumbnail_url string
}
type StreamInfo struct {
  Viewer_count int
  Game_id int
  User_name string
  Thumbnail_url string
}
type StreamAPIResult struct {
  Data []StreamInfo
  Pagination APIPagination
}
type StreamAPIResultFake struct {
  Data []StreamInfoFake
  Pagination APIPagination
}
type GameInfo struct {
  Id int           `json:",string"`
}
type GameAPIResult struct {
  Data []GameInfo
}

const ITERATIONS int = 100 // number of queries (100 each) to Helix. Rate limit is 120 per minute, so we can refresh this every 30 sec without issues.
const REFRESHTIME time.Duration = 70 * time.Second

const CLIENT_ID string = "gp762nuuoqcoxypju8c569th9wz7q5"
const OAUTH string = "Bearer fob6uh9x6z1jyv6pfrjpv0bswnaz9e"

var sorted_data []StreamInfo
var client http.Client

func apicall(path string) ([]byte, error) {
  url := "https://api.twitch.tv/helix" + path
  req, err := http.NewRequest("GET", url, nil)
  if err != nil { return []byte{}, err }
  req.Header.Add("Client-ID", CLIENT_ID)
  req.Header.Add("Authorization", OAUTH)
  
  resp, err := client.Do(req)
  if err != nil { return []byte{}, err }
  if resp.StatusCode != 200 { return []byte{}, errors.New("Non-200 response") }
  defer resp.Body.Close()
  
  return ioutil.ReadAll(resp.Body)
}

func streamapicall(path string) (StreamAPIResult, error) {
  resultbytes, err := apicall(path)
  var result StreamAPIResult
  var resultFake StreamAPIResultFake
  if err != nil { return result, err }
  err = json.Unmarshal(resultbytes, &resultFake)
  if err != nil { return result, err }
  result.Pagination = resultFake.Pagination
  for i := range resultFake.Data {
    x, err := strconv.Atoi(resultFake.Data[i].Game_id)
    if err != nil {
      x = -2
    }
    result.Data = append(result.Data, StreamInfo{
      Viewer_count: resultFake.Data[i].Viewer_count,
      Game_id: x,
      User_name: resultFake.Data[i].User_name,
      Thumbnail_url: resultFake.Data[i].Thumbnail_url,
    })
  }
  return result, nil
}
func gameapicall(path string) (GameAPIResult, error) {
  resultbytes, err := apicall(path)
  var result GameAPIResult
  if err != nil { return result, err }
  err = json.Unmarshal(resultbytes, &result)
  if err != nil { return result, err }
  return result, nil
}

func get_sorted_data() []StreamInfo {
    cur := ""

    var total_data []StreamInfo
    for i := 0; i < ITERATIONS; i++ {
        result, err := streamapicall("/streams?first=100&language=en" + cur)
        if err != nil { fmt.Println(err); break }
        total_data = append(total_data, result.Data...)
        cur = "&after=" + result.Pagination.Cursor
    }

    sort.Slice(total_data, func(i, j int) bool {
      return total_data[i].Viewer_count < total_data[j].Viewer_count
    })
    
    fmt.Println("sorted")

    return total_data
}

func get_sorted_data_loop() {
  for {
    sorted_data = get_sorted_data()
    time.Sleep(REFRESHTIME)
  }
}

var game_ids map[string]int = make(map[string]int)
func get_game_id(gamename string) int {
    gamename = strings.ToLower(strings.Trim(gamename, " \t\n"))
    if game_id, ok := game_ids[gamename]; ok {
      return game_id
    }
    result, err := gameapicall("/games?name=" + url.QueryEscape(gamename))
    var game_id int
    if err != nil {
      game_id = -1
    } else if len(result.Data) > 0 {
      game_id = result.Data[0].Id
    } else {
      game_id = -1
    }
    game_ids[gamename] = game_id
    fmt.Printf("%s %d\n", gamename, game_id)
    return game_id
}

func get_winners(nwin int, thresh int, gamename string) string {
    if nwin < 0 { nwin = 0 }
    if nwin > len(sorted_data) { nwin = len(sorted_data) }
    
    my_sorted_data := sorted_data // In case the get_sorted_data thread decides to change it, we still keep a reference to the same slice
    lastExclusive := sort.Search(len(my_sorted_data), func(i int) bool {
        return my_sorted_data[i].Viewer_count > thresh
    })
    small_streams := my_sorted_data[:lastExclusive]
    starttime := time.Now()
    
    var filtered_streams []StreamInfo
    if gamename == "all" {
      filtered_streams = make([]StreamInfo, len(small_streams))
      copy(filtered_streams, small_streams)
    } else {
      game_id := get_game_id(gamename)
      for _, stream := range small_streams {
        if stream.Game_id != 0 && stream.Game_id == game_id {
          filtered_streams = append(filtered_streams, stream)
        }
      }
    }
    fmt.Println(time.Since(starttime).Microseconds())
    
    var winners []StreamInfo
    if nwin >= len(filtered_streams) {
      winners = filtered_streams
    } else {
      for i := 0; i < nwin; i++ {
        j := i + rand.Intn(len(filtered_streams) - i)
        filtered_streams[i], filtered_streams[j] = filtered_streams[j], filtered_streams[i]
      }
      winners = filtered_streams[:nwin]
    }
    
    cards := ""
    for _, winner := range winners {
        user := winner.User_name
        url := winner.Thumbnail_url
        url = strings.ReplaceAll(url, "{width}", "400")
        url = strings.ReplaceAll(url, "{height}", "300")
        cards += fmt.Sprintf("<a target='_blank' class='card' href='https://twitch.tv/%s'><img src='%s'><div>%s</div></a>", user, url, user)
    }
    if cards == "" {
        cards = "none found :("
    }
    return cards
}


func handler(ctx *fasthttp.RequestCtx) {
  p := ctx.Path()
  fmt.Println(string(p))
  if string(p) == "/" {
    ctx.Response.Header.Set("Cache-Control", "public, max-age=7200")
    ctx.SendFile("twitchrolls.html")
    return
  }
  s1 := bytes.Index(p, []byte("/"))
  if s1 == -1 { ctx.Response.SetBodyString("Invalid input"); return }
  s2 := bytes.Index(p[s1+1:], []byte("/"))
  if s2 == -1 { ctx.Response.SetBodyString("Invalid input"); return }
  s3 := bytes.Index(p[s1+1+s2+1:], []byte("/"))
  if s3 == -1 { ctx.Response.SetBodyString("Invalid input"); return }
  s4 := bytes.Index(p[s1+1+s2+1+s3+1:], []byte("/"))
  if s4 == -1 { ctx.Response.SetBodyString("Invalid input"); return }
  s2 += s1 + 1
  s3 += s2 + 1
  s4 += s3 + 1
  nwin, err := strconv.Atoi(string(p[s2+1:s3]))
  if err != nil {
    ctx.Response.SetBodyString("Invalid input")
    return
  }
  thresh, err := strconv.Atoi(string(p[s3+1:s4]))
  if err != nil {
    ctx.Response.SetBodyString("Invalid input")
    return
  }
  gamename := string(p[s4+1:])
  fmt.Printf("%s get3 {%d} {%d} {%s}\n", time.Now().String(), nwin, thresh, gamename)
  ctx.Response.SetBodyString(get_winners(nwin, thresh, gamename))
}
func main() {
  go get_sorted_data_loop()
  fasthttp.ListenAndServe(":5000", handler)
}

File: twitchrolls.py
#!/usr/bin/env python3

import requests
import random
import itertools
import time

sorted_data = None
sorted_data_last_timestamp = None
iterations = 50 # number of queries (100 each) to Helix. Rate limit is 120 per minute, so we can refresh this every 30 sec without issues.
refreshtime = 30

client_id = 'gp762nuuoqcoxypju8c569th9wz7q5'
oauth = 'Bearer fob6uh9x6z1jyv6pfrjpv0bswnaz9e'


def get_sorted_data():
    global sorted_data
    global sorted_data_last_timestamp
    if sorted_data_last_timestamp and time.time() <= sorted_data_last_timestamp + refreshtime:
        return sorted_data

    def get_data(client_id, oauth, iterations, language="en"):
        url = "https://api.twitch.tv/helix/streams"
        HEADERS = {
            'Client-ID': client_id,
            'Authorization': oauth
        }
        PARAMS = {
            'first': 100,
            'language': 'en'
        }
        pagination = None

        total_data=[]
        for i in range(iterations):
            r = requests.get(url,headers=HEADERS, params=PARAMS)
            data = r.json()['data']
            cursor = r.json()['pagination']['cursor']
            PARAMS['after'] = cursor
            total_data.extend(data)

        return total_data

    data = get_data(client_id, oauth, iterations)
    sorted_data = sorted(data, key = lambda i: i['viewer_count'])
    sorted_data_last_timestamp = time.time()
    return sorted_data


from functools import lru_cache
@lru_cache(maxsize=None)
def get_game_id(gamename):
    HEADERS = {
        'Client-ID': client_id,
        'Authorization': oauth
    }

    PARAMS = {
        'name': gamename
    }

    game_url = "https://api.twitch.tv/helix/games"

    r = requests.get(game_url,headers=HEADERS, params=PARAMS)
    if r.status_code == 200:
      data = r.json()['data']
      print(data)
      game_id = int(data[0]['id'])
      print(game_id)
      return game_id
    else:
      return None

def get3(nwin=3, thresh=100, gamename="all"):
    gamename = gamename.lower().strip()
    sorted_data = get_sorted_data()
    small_streams = list(itertools.takewhile(lambda x: x['viewer_count'] <= thresh, sorted_data))
    if gamename != "all":
      try:
        small_streams = list(filter(lambda x: x['game_id']!='' and int(x['game_id']) == int(get_game_id(gamename)), small_streams))
      except Exception as e:
        print(e)
        return "Invalid game name, Twitch requires exact spelling"
    if len(small_streams) < nwin:
      winners = small_streams
    else:
      winners = random.sample(small_streams, k=nwin)
    cards = ""
    for winner in winners:
        user = winner['user_name']
        cards += f"<a target='_blank' class='card' href='https://twitch.tv/{user}'><img src='{winner['thumbnail_url'].format(width=400,height=300)}'><div>{user}</div></a>"
    if cards == "":
        cards = "none found"
    return cards


page = """
<script src="https://kit.fontawesome.com/813fa92790.js" crossorigin="anonymous"></script>
<style>
body {
  margin: 0; padding: 0;
  background-color: #201535;
  color: #eee;
  text-align: center;
  font-family: Candara, 'Open Sans', Arial, sans-serif;
  font-size: 1.5vw;
}
.card {
  cursor: pointer;
  width: 20%;
  display: inline-block;
  padding: 2%;
  text-align: center;
  font-weight: bold;
}
.card a{
    margin-top: 0.5em;
    text-decoration: none;
}
#rollbtn {
    background-color: #6441A4;
    color: white;
    font-weight: bold;
    padding: 0.8em;
    font-size: 2em;
    display: block;
    margin: 0.1em;
    transition: width 1s;
    overflow: hidden;
    border-width: 0.1em;
}
#rollbtn:hover {
    cursor: pointer;
}
img {
    width: 100%;
}
a, a:visited {
    color: #e0d8ef;
}
a:active {
    color: red;
}
footer {
    position: fixed;
    bottom: 2em;
    color: #999;
    text-align: center;
    width: 100%;
    opacity: 0.8;
    z-index: 100000;
    font-size: 0.8em;
}
#progress1, #progress2, #progress3 {
    display: none;
    font-size: 0.8em;
    color: #666;
}
#progress1 {
    margin-top: 2em;
}
#wrapper {
    margin-bottom: 10em;
}
input {
    background-color: transparent;
    color: inherit;
    border: solid 0.05em white;
    margin: 0.05em;
    font-size: inherit;
    font-family: inherit;
    width: 10em;
    padding: 0.1em 0.3em;
}
.w span{
    display: inline-block;
    width: 10em;
    text-align: right;
}
</style>

<div style='margin: 1em 0'>

<div style='display: inline-block; text-align: left; margin-top: 0.7em;'>
<div class='w'>
<span>Number of winners:&nbsp;</span><input type="text" id="nwin" placeholder="3" value="3">
</div><div class='w'>
<span>Max viewer count:&nbsp;</span><input type="text" id="thresh" placeholder="100" value="100">
</div><div class='w'>
<span>Game name:&nbsp;</span><input type="text" id="gamename" placeholder="all" value="all">
</div>
</div>

<div style='display: inline-block; text-align: right; vertical-align: top;'>
<button id="rollbtn"><i class="fas fa-dice"></i></button>
</div>

</div>

<div id="progress1">PROGRESS1</div>
<div id="progress2">PROGRESS2</div>
<div id="progress3">PROGRESS3</div>

<div id="wrapper"></div>

<script>
document.getElementById('rollbtn').onclick = function(e) {
    document.getElementById('progress1').style.display="none";
    document.getElementById('progress2').style.display="none";
    document.getElementById('progress3').style.display="none";
    var xhr = new XMLHttpRequest();
    xhr.open('GET', '/get3/'+parseInt(document.getElementById('nwin').value)
                   +'/'+parseInt(document.getElementById('thresh').value)
                   +'/'+document.getElementById('gamename').value, true);

    // If specified, responseType must be empty string or "text"
    xhr.responseType = 'text';

    document.getElementById('wrapper').innerHTML = "";
    document.getElementById('threshdisp').innerHTML = document.getElementById('thresh').value;
    document.getElementById('nwindisp').innerHTML = document.getElementById('nwin').value;
    document.getElementById('progress1').style.display="block";
    t1 = setTimeout(function(){document.getElementById('progress2').style.display="block";}, 3000);
    t2 = setTimeout(function(){document.getElementById('progress3').style.display="block";}, 6000);

    xhr.onload = function () {
        if (xhr.readyState === xhr.DONE) {
            if (xhr.status === 200) {
                console.log(xhr.response);
                clearTimeout(t1);
                clearTimeout(t2);
                document.getElementById('progress1').style.display="block";
                document.getElementById('progress2').style.display="block";
                document.getElementById('progress3').style.display="block";
                document.getElementById('wrapper').innerHTML = xhr.responseText;
            }
        }
    };

    xhr.send(null);
};
</script>

<footer>by <a target='_blank' href="https://twitch.tv/flar3fir3">flar3fir3</a> && <a target='_blank' href="https://clive.io">clive.io</a></footer>
""".replace("PROGRESS1", f"Getting top {iterations*100} live streams...")\
   .replace("PROGRESS2", "Filtering by viewer count &leq; <span id='threshdisp'></span>...")\
   .replace("PROGRESS3", "Randomly selecting <span id='nwindisp'></span> options...")

from flask import Flask

app = Flask(__name__)
@app.route('/')
def main():
    print("main")
    return page.replace('{nwin}', '3').replace('{thresh}', '100')

import time
from datetime import datetime
@app.route('/get3/<nwin>/<thresh>/<gamename>')
def get3_page(nwin=3, thresh=100, gamename="all"):
    print(f"{datetime.now().strftime('%H:%M:%S')} get3 {nwin} {thresh} {gamename}")
    try:
        return get3(int(nwin), int(thresh), gamename)
    except:
        return "follow flar3fir3 and clive.io pce"

import bjoern
bjoern.run(app, '0.0.0.0', 5000)
